### DISCUSSION TITLE
`ADR Suggestion` Input check in setters and flag for skipping checks for performance

### DISCUSSION THREAD

**Discussion by damskii9992** [easyscience/.github#17] [Archived]
# General
Adding checks for user inputs, and raising appropriate errors with useful error messages on wrong inputs, makes our libraries more user friendly for new users. Checks can slow down performance-sensitive code, so in order to avoid that issue we can have a global flag that skips all checks when set.

# Current implementation
Currently, user input checks are made sporadically in our code and with no capability to skip them

# Proposed implementation
An example of a setter with checks and a flag to skip them:
```
@variance.setter
@property_stack_deco
def variance(self, variance_float: float) -> None:
    """
    Set the variance.

    :param variance_float: Variance as a float
    """
    if not UNSAFE:
        if not isinstance(variance_float, numbers.Number):
            raise TypeError(f'{variance_float=} must be a number')
        if variance_float < 0:
            raise ValueError(f'{variance_float=} must be positive')
    self._scalar.variance = variance_float
```
where UNSAFE is a global boolean which can, for example, be set by our calculators and minimizers.

**Comment by rozyczko** [comment]
```if not UNSAFE:``` is quite unintuitive.
Wouldn't it be better to just have ```if SAFE:```

**Reply by AndrewSazonov** [reply]
I would vote for `if SAFE`. It's way easier to read.

**Comment by damskii9992** [comment]
Just checking in, but was this ADR suggestion actually agreed upon?

**Reply by rozyczko** [reply]
We probably need to assess the suggestion in terms of actual performance effect. We will already be checking one conditional, so if the check involves only a type or max/min, maybe there's no reason to make the code more complex?

**Reply by damskii9992** [reply]
I suppose it would be easy to implement skipping checks in setters in the future if it was found to meaningfully degrade performance.
So maybe this ADR should simply be about having input checks in setters?

**Reply by rozyczko** [reply]
Yes, but then it isn't an ADR material but a coding style issue, which we should cover in our `Coding Style` guide.

**Reply by damskii9992** [reply]
Do we have a `Coding Style` guide? And if yes, where to find it?

**Comment by rozyczko** [comment]
The author requested closing the discussion.
